### PR TITLE
feat: validate page path

### PR DESCRIPTION
**Pull request by TrySound**
Here added description tooltip and additional validation to constrain UrlPattern syntax to fit most framework abilities.

Screenshot 2023-12-13 at 15 58 11
Screenshot 2023-12-13 at 15 56 45

## Code Review

- [ ] hi @kof, I need you to do
  - conceptual review (architecture, feature-correctness)
  - detailed review (read every line)
  - test it on preview

## Before requesting a review

- [ ] made a self-review
- [ ] added inline comments where things may be not obvious (the "why", not "what")

## Before merging

- [ ] tested locally and on preview environment (preview dev login: 5de6)
- [ ] updated [test cases](https://github.com/webstudio-is/webstudio/blob/main/apps/builder/docs/test-cases.md) document
- [ ] added tests
- [ ] if any new env variables are added, added them to `.env.example` and the `builder/env-check.js` if mandatory

### CODE REVIEW

**Comment by kof** [comment]
- [ ] Path tooltip would look better above instead of on the right

File: apps/builder/app/builder/features/sidebar-left/panels/pages/url-pattern.ts [hunk 1]
```diff
@@ -38,3 +37,77 @@ export const compilePathnamePattern = (
...
+  // show segment errors only when syntax is valid
+  if (messages.length > 0) {
+    return messages;
+  }
+
+  for (const segment of pathname.split("/")) {
+    const group = segment.match(/(?<group>:\w+(\*|\?)?)/)?.groups?.group;
+    if (group) {
+      if (group.length !== segment.length) {
+        messages.push(`Cannot use named group at '${segment}'`);
```
Comment (kof) [inline apps/builder/app/builder/features/sidebar-left/panels/pages/url-pattern.ts:103]: message seems to have an grammar issue

File: apps/builder/app/builder/features/sidebar-left/panels/pages/url-pattern.ts [hunk 2]
```diff
@@ -38,3 +37,77 @@ export const compilePathnamePattern = (
...
+      }
+    } else if (segment.includes("*")) {
+      if (segment.length > 1) {
+        messages.push(`Cannot use wildcard at '${segment}'`);
```
Comment (kof) [inline apps/builder/app/builder/features/sidebar-left/panels/pages/url-pattern.ts:107]: message seems to have an grammar issue

File: apps/builder/app/builder/features/sidebar-left/panels/pages/url-pattern.ts [hunk 3]
```diff
@@ -38,3 +37,77 @@ export const compilePathnamePattern = (
...
+
+  // * in the middle
+  const wildcardGroups = Array.from(
+    // skip matching with :name before *
+    // skip matching with ? after *
+    // skip matching with * in the end
+    pathname.matchAll(/(?<!:\w+)\*(?!\?)(?!$)/g)
+  ).flat();
+  if (wildcardGroups.length > 0) {
+    messages.push(`Wildcard group '*' is allowed only in the end`);
```
Comment (kof) [inline apps/builder/app/builder/features/sidebar-left/panels/pages/url-pattern.ts:91]: message seems to have an grammar issue

**Review by kof — COMMENTED** [review]
please check all messages for grammar correctness

**Review by kof — APPROVED** [review]

**Comment by TrySound** [comment]
>  Path tooltip would look better above instead of on the right

All validation here works this way

File: apps/builder/app/builder/features/sidebar-left/panels/pages/settings.tsx [hunk 4]
```diff
@@ -171,13 +178,16 @@ const validateValues = (
...
   }
-  if (
-    pages !== undefined &&
-    values.path !== undefined &&
-    isPathUnique(pages, pageId, values.path) === false
-  ) {
-    errors.path = errors.path ?? [];
-    errors.path.push("All paths must be unique");
+  if (pages !== undefined && values.path !== undefined) {
+    if (isPathUnique(pages, pageId, values.path) === false) {
```
[the remark this answers is not in the document]
Comment (TrySound) [reply apps/builder/app/builder/features/sidebar-left/panels/pages/settings.tsx:182]: This is old code. Just moved a bit.

**Pull request merged**